File: Leetcode/92-Reverse-Linked-List-II/ReverseLinkedListII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if head == nil || n <= m {
		return head
	}
	var dummy *ListNode
	dummy = &ListNode{Val: 0, Next: head}
	cur := dummy
	i := 0
	for i+1 < m {
		if cur != nil {
			cur = cur.Next
		}
		i++
	}
	prev := cur
	var n1 *ListNode
	if prev != nil {
		n1 = prev.Next
	}
	var n2 *ListNode
	if n1 != nil {
		n2 = n1.Next
	}
	var ran = n - m
	for ran > 0 && n2 != nil {
		n2.Next, n1, n2 = n1, n2, n2.Next
		ran--
	}
	if prev != nil && prev.Next != nil {
		prev.Next.Next = n2
		prev.Next = n1
	}
	return dummy.Next
}

File: Lintcode/36-Reverse-Linked-List-II/ReverseLinkedListII.java

/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class ReverseLinkedListII {
	/**
     * @param ListNode head is the head of the linked list 
     * @oaram m and n
     * @return: The head of the reversed ListNode
     */
	public ListNode reverseBetween(ListNode head, int m, int n) {
		if (head == null || n <= m) {
			return head;
		}

		ListNode mDummy = new ListNode(0);
		mDummy.next = head;
		ListNode cur = mDummy;
		for (int i = 0; i + 1 < m; i++) {
			if (cur != null) {
				cur = cur.next;
			}
		}
		ListNode mPrev = cur;
		ListNode n1 = null;
		if (mPrev != null) {
			n1 = mPrev.next;
		}
		ListNode n2 = null;
		if (n1 != null) {
			n2 = n1.next;
		}
		int range = n - m;
		while (range > 0 && n2 != null) {
			ListNode temp = n2.next;
			n2.next = n1;
			n1 = n2;
			n2 = temp;
			range--;
		}
		if (mPrev != null) {
			mPrev.next.next = n2;
			mPrev.next = n1;
		}
		return mDummy.next;
	}
}
